classes/merge_request.php

```diff
@@ -0,0 +1,204 @@
+        if (isset($this->data->{$name})) {
+            if ($name == 'log') {
+                $value = json_encode($value);
+            }
+            $this->data->{$name} = $value;
+        }
+    }
+
+    /**
+     * Convert log to the new format.
+     * @param log field;
+     * @return
+     */
+    public function convert_log_to_new_format() {
+        if (!isset($this->data->log) || empty($this->data->log)) {
+            return;
+        }
+        $logs = [];
+        $logs[1] = json_decode($this->data->log, true);
+        $this->data->log = json_encode($logs);
+    }
+
+    /**
+     * Get funtcion.
+     * @return data
+     */
+    public function get_record(): \stdClass {
+        return $this->data;
+    }
+
+    /**
+     * Export data to new merge users table.
+     */
+    public static function export_data_to_new_table(): void {
+        global $DB;
+        $sort = "id ASC";
+        $records = $DB->get_recordset(self::TABLE_MERGE_REQUEST_OLD,
+                                        null,
+                                        $sort);
+        if (!$records) {
+            // There is no need to migrate. That's all!
+            return;
+        }
+        $oldrequests = [];
+        $orderedoldrequests = [];
+        foreach ($records as $item) {
+            if (!isset($oldrequests[$item->fromuserid])) {
+                $oldrequests[$item->fromuserid] = [];
+            }
+            if (isset($oldrequests[$item->fromuserid][$item->touserid])) {
+                 $baseitem = $oldrequests[$item->fromuserid][$item->touserid];
+            } else {
+                $baseitem = new \stdClass();
+                $baseitem->removeuserfield = 'id';
+                $baseitem->removeuservalue = $item->fromuserid;
+                $baseitem->removeuserid = $item->fromuserid;
+                $baseitem->keepuserfield = 'id';
+                $baseitem->keepuservalue = $item->touserid;
+                $baseitem->keepuserid = $item->touserid;
+                if (isset( $item->mergedbyuserid)) {
+                    $baseitem->mergedbyuserid = $item->mergedbyuserid;
+                }
+                $baseitem->timeadded = $item->timemodified;
+                $baseitem->log = [];
+                $oldrequests[$item->fromuserid][$item->touserid] = $baseitem;
+                $orderedoldrequests[$baseitem->timeadded] = $baseitem;
+            }
+            // Old logs are obtained in ASC order, so we can safely update this values.
+            $baseitem->timemodified = $item->timemodified;
+            $baseitem->timecompleted = $item->timemodified;
+            // Append logs to the list.
+            $baseitem->log[$item->timemodified] = json_decode($item->log, false);
+            // $baseitem->log[0] = "Migrated from old record with id = ".$item->id;.
+            $baseitem->status = ($item->status == 1) ? self::COMPLETED_WITH_SUCCESS : self::COMPLETED_WITH_ERRORS;
+            $baseitem->retries = 0;
```

Comment:
I think here we should have this sentence instead, to reflect how many times the pair of users were merged in the old version:

```php
$baseitem->retries = count($baseitem->log);
```

This way, if in the old version of the plugin, the pair of users were tried 5 times, it will show `5`, instead of 0. In fact, at least this value should be 1, since at least there would be one record in the old table.

Reply:
In addition, related to the `retries`: this attribute is related to the concept of number of trials to merge two users. However, there is no attribute meaning how many sucessful mergings were produced. I think we should have this additional attribute, named for instance `successfullretries`. In the best scenario, there should be 2 `retries` and 2 `successfullretries`, meaning that on the first attempt, the merge was ok, and it was queued again to ensure all potential pending data was merged.

Having said that, I think we should reconsider how we deal with `retries`, because if we have 3 retries, and the last was ok, the additional one to cover those potential pending data will not be processed by no means. But it should. Mmmmm... we have to think about it.

Reply:
I changed $baseitem->retries = count($baseitem->log);